src/source/view/controls/TextInput.bs

```diff
@@ -0,0 +1,207 @@
+      m.inputImage = m.createSGNode("mv_Poster", m.top, "inputImage")
+      m.backgroundColor = m.createSGNode("mv_Rectangle", m.top, "backgroundColor")
+      m.divider = m.createSGNode("mv_Rectangle", m.top, "divider")
+
+      m.label = m.createSGNode("mv_Label", m.top, "label")
+      m.errorLabel = m.createSGNode("mv_Label", m.top, "errorLabel")
+      m.subTextLabel = m.createSGNode("mv_Label", m.top, "subTextLabel")
+
+      m.errorBackgroundPoster = m.createSGNode("mv_Poster", m.top, "errorBackgroundPoster")
+    end function
+
+    '+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    '++ callbacks
+    '+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+    function onStyleKeyChange(key as string)
+      m.style = m.styleManager@.getStyle(key)
+      m.redraw()
+    end function
+
+    private function redraw() as void
+      if m.style = invalid
+        m.style = m.styleManager@.getStyle(m.styleKey)
+      end if
+
+      'hack to allow this view to work in xml
+      if m.label = invalid
+        m.createViews()
+      end if
+
+      m.applyState(m.isFocused, m.isDisabled, m.hasError)
+
+      if m.text <> invalid and m.text.trim() <> ""
+        m.label.text = m.text
+      else
+        m.label.text = m.hintText
+      end if
+
+      m.divider.translation = [m.label.translation[0], m.label.translation[1] + m.label.boundingRect()["height"] + 10]
+      m.inputImage.translation = [m.divider.translation[0] + m.divider.width + m.inputImage.translation[0], m.divider.translation[1] - m.divider.size[1] / 2 - m.inputImage.size[1] / 2]
+
+      m.subTextLabel.text = m.subText
+      m.subTextLabel.width = m.divider.size[0]
+      m.subTextLabel.translation = [m.divider.translation[0], m.divider.translation[1] + m.divider.size[1] + 10]
+      m.subTextLabel.wrap = true
+
+      m.errorLabel.text = m.errorText
+      m.errorLabel.translation = [m.inputImage.translation[0] + m.inputImage.size[0] * 2,  m.inputImage.translation[1] + m.inputImage.size[1] / 2 - m.errorLabel.boundingRect()["height"] / 2]
+      m.errorLabel.visible = m.errorText <> invalid and m.errorText.trim() <> "" and m.hasError
+
+      if m.backgroundColor.size[0] > 0:
+        m.divider.visible = false
+        m.label.translation = [m.backgroundColor.translation[0] + 10, m.backgroundColor.translation[1] + 10]
+        m.subTextLabel.translation = [m.label.translation[0], m.backgroundColor.translation[1] + m.backgroundColor.size[1] + 10]
+        m.inputImage.translation = [m.backgroundColor.translation[0] + m.backgroundColor.size[0] + 10, m.backgroundColor.translation[1] + m.backgroundColor.size[1] / 2 - m.inputImage.size[1] / 2 ]
+        m.errorLabel.translation =  [m.inputImage.translation[0] + m.inputImage.size[0] + 10, m.inputImage.translation[1] + m.inputImage.size[1] / 2 - m.errorLabel.boundingRect()["height"] / 2]
+      end if
+    end function
+
+    ' this is just a POC, validators should have their own class
```

Comment:
good idea; but let's make instead a ValidationUtil class - we can then use it in the TexTinputScreen - that way we can use this as a simple dumb control, without having to change the implementation to add new validations.

I prefer composition over inhreitence for this kind of thing: https://betterprogramming.pub/prefer-composition-over-inheritance-1602d5149ea1

I also prefer SRP https://stackify.com/solid-design-principles/ - so :
 - the textinput Just hanldes textinput, 
  - the ValidatinUtil class (or validation.util namespace) just provides validatoin functions,
  -  the TextInputScreen's only responsability is to control the elements on screen and orchestrate their focus and validattion via the textinput controls and validation util